storage/src/merkle_storage.rs

```diff
@@ -717,170 +730,99 @@ impl MerkleStorage {
+        };
+        self.compute_new_root_with_change(root, &key, Some(new_node))
+    }
+
     /// Delete an item from the staging area.
-    pub fn delete(&mut self, key: &ContextKey) -> Result<(), MerkleError> {
+    pub fn delete(&mut self, new_tree_id: TreeId, key: &ContextKey) -> Result<(), MerkleError> {
         let instant = Instant::now();
-        let act = Arc::make_mut(&mut self.actions);
-        // store action
-        act.push(Action::Remove(RemoveAction { key: key.to_vec() }));
+        let root = self.get_staged_root();
+        let new_root_hash = &self._delete(&root, key)?;
+        self.set_stage_root(&self.get_tree(new_root_hash)?, new_tree_id);
         self.update_execution_stats("Delete".to_string(), Some(&key), &instant);
-        Ok(())
-    }
 
-    /// Copy subtree under a new path.
-    /// TODO Consider copying values!
-    pub fn copy(&mut self, from_key: &ContextKey, to_key: &ContextKey) -> Result<(), MerkleError> {
-        let instant = Instant::now();
-        let act = Arc::make_mut(&mut self.actions);
-        // store action
-        act.push(Action::Copy(CopyAction {
-            from_key: from_key.to_vec(),
-            to_key: to_key.to_vec(),
-        }));
-        // TODO: do we need to include from_key in stats?
-        self.update_execution_stats("CopyToDiff".to_string(), Some(&to_key), &instant);
         Ok(())
     }
 
-    fn add_empty_tree_to_staging(&mut self) -> Result<Option<usize>, MerkleError> {
-        let tree = Tree::new();
-        let hash = hash_tree(&tree)?;
-        self.put_to_staging_area(&hash, Entry::Tree(tree))
-    }
-
-    /// If current staging tree does not exist yet, create a new empty tree
-    fn ensure_stage_tree_exists(&mut self) -> Result<(), MerkleError> {
-        match &self.current_stage_tree {
-            None => {
-                let tree = Tree::new();
-                self.current_stage_tree = Some(tree.clone());
-                let hash = hash_tree(&tree)?;
-                self.current_stage_tree_hash = Some(hash);
-                self.put_to_staging_area(&hash, Entry::Tree(tree))?;
-            }
-            Some(_tree) => (),
+    fn _delete(&mut self, root: &Tree, key: &ContextKey) -> Result<EntryHash, MerkleError> {
+        if key.is_empty() {
+            return Ok(hash_tree(root)?);
         }
-        Ok(())
+        self.compute_new_root_with_change(root, &key, None)
     }
 
-    fn increase_refcnt_for_staging_entry(&mut self, hash: &EntryHash) -> Result<(), MerkleError> {
-        let idx = self.staged_get_idx(hash);
-        match idx {
-            Some(idx) => {
-                self.staged[idx].1 += 1;
-            }
-            None => {
-                return Err(MerkleError::EntryNotFoundInStaging {
-                    hash: HashType::ContextHash.hash_to_b58check(hash)?,
-                });
-            }
-        }
+    /// Copy subtree under a new path.
+    /// TODO Consider copying values!
```

Comment:
@mat13mn what is this about?

Reply:
seems like really old comment 
https://github.com/simplestaking/tezedge/commit/430398ba6997b4dce9d95c577b594efa59bf3d02#diff-8e32c1599e2683630b3d12ff50bfbc818afbd7ac8205c662925ba5dfd1b1743eR239

probably can be removed

Reply:
Ok cool, that explains why I couldn't make sense of it, will remove.